crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -2843,7 +2843,34 @@ impl Ord for Fork {
         // work for the latter, but the inverse is unlikely to be true.
         let self_bound = requires_python_marker(&self.markers).unwrap_or_default();
         let other_bound = requires_python_marker(&other.markers).unwrap_or_default();
-        other_bound.cmp(&self_bound)
+
+        other_bound.cmp(&self_bound).then_with(|| {
+            // If there's no difference, prioritize forks with upper bounds. We'd prefer to solve
+            // `numpy <= 2` before solving `numpy >= 1`, since the resolution produced by the former
+            // might work for the latter, but the inverse is unlikely to be true due to maximum
+            // version selection. (Selecting `numpy==2.0.0` would satisfy both forks, but selecting
+            // the latest `numpy` would not.)
+            let self_upper_bounds = self
+                .dependencies
+                .iter()
+                .filter(|dep| {
+                    dep.version
+                        .bounding_range()
+                        .is_some_and(|(_, upper)| !matches!(upper, Bound::Unbounded))
+                })
+                .count();
+            let other_upper_bounds = other
+                .dependencies
+                .iter()
+                .filter(|dep| {
+                    dep.version
+                        .bounding_range()
+                        .is_some_and(|(_, upper)| !matches!(upper, Bound::Unbounded))
+                })
+                .count();
+
+            self_upper_bounds.cmp(&other_upper_bounds)
```

Comment:
Just a heuristic of course, and we don't really have a ton of examples to inform this work.